File: packages/sdk/src/bans.ts
import type {
  Ban,
  BanHistoryEntry,
  GameId,
  GroupId,
  Page,
  PageOptions,
  UserId,
} from "@junjo/shared";
import { JunjoError } from "./errors.js";
import type { HttpClient } from "./http.js";
import { paginate } from "./pagination.js";

interface WireGameBan {
  id: string;
  gameId: string;
  userId: string;
  bannedAt: string;
  expiresAt: string | null;
  reason: string | null;
  bannedBy: string | null;
}

function deserializeBan(w: WireGameBan): Ban {
  return {
    id: w.id,
    gameId: w.gameId as GameId,
    userId: w.userId as UserId,
    bannedAt: new Date(w.bannedAt),
    expiresAt: w.expiresAt === null ? null : new Date(w.expiresAt),
    reason: w.reason,
    bannedBy: w.bannedBy === null ? null : (w.bannedBy as UserId),
  };
}

export interface WireBanHistoryEntry {
  id: string;
  gameId: string;
  userId: string;
  scope: "game" | "group";
  groupId: string | null;
  kind: "set" | "lifted";
  reason: string | null;
  expiresAt: string | null;
  eventAt: string;
  actorUserId: string | null;
}

export function deserializeBanHistoryEntry(w: WireBanHistoryEntry): BanHistoryEntry {
  return {
    id: w.id,
    gameId: w.gameId as GameId,
    userId: w.userId as UserId,
    scope: w.scope,
    groupId: w.groupId === null ? null : (w.groupId as GroupId),
    kind: w.kind,
    reason: w.reason,
    expiresAt: w.expiresAt === null ? null : new Date(w.expiresAt),
    eventAt: new Date(w.eventAt),
    actorUserId: w.actorUserId === null ? null : (w.actorUserId as UserId),
  };
}

export interface CreateBanInput {
  // External user id of the user to ban (Clerk sub, Supabase uuid,
  // Roblox UserId-as-string -- whatever the dev's auth provider returns).
  userId: UserId;
  // Optional human-readable reason; surfaces in audit / dashboard.
  reason?: string | null;
  // Optional ISO timestamp / Date for time-bounded bans. Omit / null =
  // permanent. Lazy expiry on read; the server does not auto-clean
  // expired rows.
  expiresAt?: Date | string | null;
  // Optional moderator attribution. The dev's external user id of the
  // operator pressing the ban button. Auto-creates a JunjoUser if the
  // actor hasn't been seen in this game (mirrors the target user).
  // Surfaces back as `Ban.bannedBy` and on audit / BanHistory rows.
  actorUserId?: UserId;
}

export interface ListBansOptions extends PageOptions {
  // Default false. When true, also returns rows whose `expiresAt` is in
  // the past (the runtime ban-check ignores those, but operators may
  // want to see them in the dashboard).
  includeExpired?: boolean;
}

export interface ListBanHistoryOptions extends PageOptions {
  // Filter to one ban surface. Omit for both. Forced to "group" when
  // `groupId` is supplied; supplying both with `scope: "game"` is a
  // 400 error.
  scope?: "game" | "group";
  groupId?: GroupId;
}

// Game-level bans. Per-group bans live on `MembersApi` / `GroupsApi`
// alongside kick semantics.
//
// Game-level ban applies across every group in the game: the user
// cannot accept invitations or open-join any group while the ban is
// active. Per-group ban (`groups.ban(...)`) is scoped to one group.
// The two compose -- enforcement on the server checks game-level
// first, then per-group.
export class BansApi {
  constructor(private readonly http: HttpClient) {}

  async list(opts?: ListBansOptions): Promise<Page<Ban>> {
    const params = new URLSearchParams();
    if (opts?.limit !== undefined) params.set("limit", String(opts.limit));
    if (opts?.cursor !== undefined) params.set("cursor", opts.cursor);
    if (opts?.includeExpired === true) params.set("includeExpired", "true");
    const qs = params.toString();
    const path = qs ? `/v1/bans?${qs}` : "/v1/bans";
    const wire = await this.http.get<{ items: WireGameBan[]; nextCursor: string | null }>(path);
    return {
      items: wire.items.map(deserializeBan),
      nextCursor: wire.nextCursor,
    };
  }

  // Async-iterator wrapper over `list(...)`. See GroupsApi.listAll.
  listAll(opts?: { limit?: number; includeExpired?: boolean }): AsyncGenerator<Ban> {
    return paginate((cursor) => this.list({ ...opts, cursor }));
  }

  async add(input: CreateBanInput): Promise<Ban> {
    const body: Record<string, unknown> = { userId: input.userId };
    if (input.reason !== undefined) body.reason = input.reason;
    if (input.expiresAt !== undefined) {
      body.expiresAt =
        input.expiresAt === null
          ? null
          : input.expiresAt instanceof Date
            ? input.expiresAt.toISOString()
            : input.expiresAt;
    }
    if (input.actorUserId !== undefined) body.actorUserId = input.actorUserId;
    const wire = await this.http.post<WireGameBan>("/v1/bans", body);
    return deserializeBan(wire);
  }

  async remove(userId: UserId, opts?: { actorUserId?: UserId }): Promise<void> {
    await this.http.delete<unknown>(
      `/v1/bans/${encodeURIComponent(userId)}`,
      opts?.actorUserId !== undefined ? { actorUserId: opts.actorUserId } : undefined,
    );
  }

  // Fetch the current active game-level ban for a user. Returns null
  // if the user is not banned, the ban has expired, or the user has
  // never been seen in this game.
  async get(userId: UserId): Promise<Ban | null> {
    try {
      const wire = await this.http.get<WireGameBan>(`/v1/bans/${encodeURIComponent(userId)}`);
      return deserializeBan(wire);
    } catch (err) {
      if (err instanceof JunjoError && err.code === "not_found") return null;
      throw err;
    }
  }

  // Append-only ban-event timeline for a user in this game. Includes
  // both game-scope and group-scope rows by default. Cursor-paginated,
  // newest-first.
  async history(userId: UserId, opts?: ListBanHistoryOptions): Promise<Page<BanHistoryEntry>> {
    const params = new URLSearchParams();
    if (opts?.limit !== undefined) params.set("limit", String(opts.limit));
    if (opts?.cursor !== undefined) params.set("cursor", opts.cursor);
    if (opts?.scope !== undefined) params.set("scope", opts.scope);
    if (opts?.groupId !== undefined) params.set("groupId", opts.groupId);
    const qs = params.toString();
    const path = qs
      ? `/v1/bans/${encodeURIComponent(userId)}/history?${qs}`
      : `/v1/bans/${encodeURIComponent(userId)}/history`;
    const wire = await this.http.get<{
      items: WireBanHistoryEntry[];
      nextCursor: string | null;
    }>(path);
    return {
      items: wire.items.map(deserializeBanHistoryEntry),
      nextCursor: wire.nextCursor,
    };
  }

  // Async-iterator wrapper over `history(...)`. Walks every page of
  // ban events for the user until exhausted.
  historyAll(
    userId: UserId,
    opts?: { limit?: number; scope?: "game" | "group"; groupId?: GroupId },
  ): AsyncGenerator<BanHistoryEntry> {
    return paginate((cursor) => this.history(userId, { ...opts, cursor }));
  }
}

File: packages/sdk-roblox/src/bans.lua
--!nonstrict
-- Nonstrict, not strict: cross-module `require(script.Parent.X)` types
-- and the metatable-OOP idiom below need the Roblox definition files to
-- pass strict analysis, which CI cannot run yet. Public signatures
-- carry annotations regardless.
--
-- Game-level bans namespace. Mirrors the Junjo.io TypeScript SDK's
-- `junjo.bans` surface (add / remove / get / list / history plus the
-- listAll / historyAll iterators). Per-group bans live on the groups
-- namespace (`groups:ban` / `groups:unban`) alongside kick semantics;
-- the two compose -- server-side enforcement checks game-level first,
-- then per-group.
--
-- Roblox game servers are TRUSTED callers: they hold the per-game
-- `jk_` key, so the full moderation surface is legitimately available
-- server-side. Never ship the key to clients.
--
-- Every method returns the parsed server response verbatim. Timestamp
-- fields stay as ISO 8601 strings (convert via `DateTime.fromIsoDate`
-- when a DateTime value is wanted).

local tryGet = require(script.Parent.TryGet)
local pageAll = require(script.Parent.PageAll)
local Types = require(script.Parent.Types)

type GameBan = Types.GameBan
type BanHistoryEntry = Types.BanHistoryEntry
type Page<T> = Types.Page<T>

local Bans = {}
Bans.__index = Bans

-- Shallow-copies opts so a pageAll loop can vary the cursor without
-- mutating the caller's table.
local function withCursor(opts, cursor: string?)
	local merged = {}
	if opts then
		for k, v in pairs(opts) do
			merged[k] = v
		end
	end
	merged.cursor = cursor
	return merged
end

function Bans.new(http)
	local self = setmetatable({}, Bans)
	self._http = http
	return self
end

-- Game-wide ban. Idempotent on a still-active ban for the same user
-- (the server returns the existing row); an expired ban is replaced by
-- a fresh one. `opts.expiresAt` (an ISO 8601 string) enables
-- time-bounded bans; omit it for a permanent ban. `opts.actorUserId`
-- attributes the action to a specific moderator in audit + history.
function Bans:add(
	userId: string,
	opts: { reason: string?, expiresAt: string?, actorUserId: string? }?
): GameBan
	local body = { userId = userId }
	if opts then
		if opts.reason ~= nil then body.reason = opts.reason end
		if opts.expiresAt ~= nil then body.expiresAt = opts.expiresAt end
		if opts.actorUserId ~= nil then body.actorUserId = opts.actorUserId end
	end
	return self._http:post("/v1/bans", body)
end

-- Lift a game-wide ban. The body is genuinely optional on the server
-- (no actor = null actor), so the DELETE goes out body-less unless
-- `opts.actorUserId` is supplied. A missing ban raises a JunjoError
-- with code = "not_found".
function Bans:remove(userId: string, opts: { actorUserId: string? }?)
	local body = nil
	if opts and opts.actorUserId ~= nil then
		body = { actorUserId = opts.actorUserId }
	end
	self._http:delete("/v1/bans/" .. self._http:encode(userId), body)
end

-- Current active game-level ban for the user, or nil when the user is
-- not banned, the ban has expired, or the user has never been seen in
-- this game (all surface as 404 on the server).
function Bans:get(userId: string): GameBan?
	return tryGet(self._http, "/v1/bans/" .. self._http:encode(userId))
end

-- Active bans, newest-first, cursor-paginated. Pass
-- `opts.includeExpired = true` to also return rows whose `expiresAt`
-- is in the past (runtime ban checks ignore those, but operators may
-- want to see them).
function Bans:list(
	opts: { limit: number?, cursor: string?, includeExpired: boolean? }?
): Page<GameBan>
	local query = {}
	if opts and opts.limit ~= nil then
		table.insert(query, "limit=" .. self._http:encode(opts.limit))
	end
	if opts and opts.cursor ~= nil then
		table.insert(query, "cursor=" .. self._http:encode(opts.cursor))
	end
	if opts and opts.includeExpired == true then
		table.insert(query, "includeExpired=true")
	end
	local path = "/v1/bans"
	if #query > 0 then
		path = path .. "?" .. table.concat(query, "&")
	end
	return self._http:get(path)
end

-- Generic-for iterator over `list(...)` that walks every page until
-- `nextCursor` is nil. `opts.limit` is the per-page size hint
-- (server-capped); `opts.cursor` is owned by the iterator and ignored
-- if supplied.
function Bans:listAll(opts: { limit: number?, includeExpired: boolean? }?): () -> GameBan?
	return pageAll(function(cursor)
		return self:list(withCursor(opts, cursor))
	end)
end

-- Append-only ban-event timeline for one user in this game,
-- newest-first, cursor-paginated. Includes both game-scope and
-- group-scope rows by default; `opts.scope` filters to one surface and
-- `opts.groupId` narrows to one group (the server forces scope=group
-- in that case, and rejects an explicit scope="game" + groupId combo).
function Bans:history(
	userId: string,
	opts: { limit: number?, cursor: string?, scope: string?, groupId: string? }?
): Page<BanHistoryEntry>
	local query = {}
	if opts and opts.limit ~= nil then
		table.insert(query, "limit=" .. self._http:encode(opts.limit))
	end
	if opts and opts.cursor ~= nil then
		table.insert(query, "cursor=" .. self._http:encode(opts.cursor))
	end
	if opts and opts.scope ~= nil then
		table.insert(query, "scope=" .. self._http:encode(opts.scope))
	end
	if opts and opts.groupId ~= nil then
		table.insert(query, "groupId=" .. self._http:encode(opts.groupId))
	end
	local path = "/v1/bans/" .. self._http:encode(userId) .. "/history"
	if #query > 0 then
		path = path .. "?" .. table.concat(query, "&")
	end
	return self._http:get(path)
end

-- Generic-for iterator over `history(...)`. See `listAll` for the
-- iteration contract.
function Bans:historyAll(
	userId: string,
	opts: { limit: number?, scope: string?, groupId: string? }?
): () -> BanHistoryEntry?
	return pageAll(function(cursor)
		return self:history(userId, withCursor(opts, cursor))
	end)
end

return Bans
